python/ray/data/_internal/stage_impl.py

```diff
@@ -147,50 +150,90 @@ class ZipStage(AllToAllStage):
+            # NOTE: _split_at_indices has a no-op fastpath if the blocks are already
+            # aligned.
+            aligned_other_blocks_with_metadata = _split_at_indices(
+                other_blocks_with_metadata,
+                indices,
+                other_block_list._owned_by_consumer,
+                other_block_rows,
+            )
+            del other_blocks_with_metadata
 
+            blocks = [b for b, _ in blocks_with_metadata]
+            other_blocks = aligned_other_blocks_with_metadata[0]
+            del blocks_with_metadata, aligned_other_blocks_with_metadata
             if clear_input_blocks:
                 block_list.clear()
+                other_block_list.clear()
 
-            if len(blocks1) != len(blocks2):
-                # TODO(ekl) consider supporting if num_rows are equal.
-                raise ValueError(
-                    "Cannot zip dataset of different num blocks: {} vs {}".format(
-                        len(blocks1), len(blocks2)
-                    )
-                )
-
-            def do_zip(block1: Block, block2: Block) -> (Block, BlockMetadata):
-                stats = BlockExecStats.builder()
-                b1 = BlockAccessor.for_block(block1)
-                result = b1.zip(block2)
-                br = BlockAccessor.for_block(result)
-                return result, br.get_metadata(input_files=[], exec_stats=stats.build())
+            do_zip = cached_remote_fn(_do_zip, num_returns=2)
 
-            do_zip_fn = cached_remote_fn(do_zip, num_returns=2)
-
-            blocks = []
-            metadata = []
-            for b1, b2 in zip(blocks1, blocks2):
-                res, meta = do_zip_fn.remote(b1, b2)
-                blocks.append(res)
-                metadata.append(meta)
+            out_blocks = []
+            out_metadata = []
+            for block, other_blocks in zip(blocks, other_blocks):
+                # For each block in block_list, zip it together with 1 or more blocks
+                # from other. We're guaranteed to have that block and other_blocks have
+                # the same number of rows.
+                res, meta = do_zip.remote(block, *other_blocks)
+                out_blocks.append(res)
+                out_metadata.append(meta)
 
             # Early release memory.
-            del blocks1, blocks2
+            del blocks, other_blocks
 
             # TODO(ekl) it might be nice to have a progress bar here.
-            metadata = ray.get(metadata)
+            out_metadata = ray.get(out_metadata)
             blocks = BlockList(
-                blocks, metadata, owned_by_consumer=block_list._owned_by_consumer
+                out_blocks,
+                out_metadata,
+                owned_by_consumer=block_list._owned_by_consumer,
             )
             return blocks, {}
 
         super().__init__("zip", None, do_zip_all)
 
 
+def _do_zip(block: Block, *other_blocks: Block) -> (Block, BlockMetadata):
```

Comment:
Annotate other_blocks as List[Block] for readability?

Reply:
Variadic args should have the type annotation of a single one of the arguments, not the collection. https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values

Reply:
It's List[ObjectRef[Block] from splitting output, any reason this to be a variadic arg?

Reply:
This is used as a Ray task function, and Ray only resolves object refs that are top-level arguments, so we want to have each of these data blocks as top-level arguments to (1) get automatic materialization, (2) ensure the task isn't scheduled until all blocks are resolved, and (3) take advantage of locality-aware scheduling of the task; we wouldn't get any of those 3 things if we did a `ray.get()` in the task function, if that's what you're recommending.

Reply:
ok, makes sense!